src/app/application/services/data-source/DataSourcePool.ts

```diff
@@ -0,0 +1,80 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const create = (src: string, router: Router<Source>): CallableEventSource => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
+    }
+  })
+  return _source
+}
+const destroy = (_src: string, source: CallableEventSource) => {
+  if (source) {
+    source.close()
+  }
+}
```

Comment:
_src is not used. Why do we have this method should we just do the suggestion below.

Reply:
I suppose it was just because "as `create` is separate, lets make `destroy` separate also".

---

Ah, I think also I would really like to consider getting shared pool instantiation out of the constructor, in fact its more I would really like to get both `create` and `destroy` functions out of this file entirely, so you can define them externally to DataSource and pass them in if you want to. Right now I'm restricted by the `router` also being instantiated in the constructor and I don't want to have to pass the `router`, you can see I need that in the `create` function.

🤔 actually what I want to do here is a bit difficult to explain here, might be better to explain next time we pair on this.

For the moment I'm going to do the suggestion you added here, but would be good to get your input on what I would like to do here, right now what I would like to do isn't urgent at all though.